server/actions/ssh/runScript.js

```diff
@@ -1,101 +1,111 @@
-  const { repo, input, path, bonusDate, name } = assignment;
-  const repoPath = `.private_repo/${path.match(/.*\/(.+?)\/?$/)[1]}`;
-
-  return command({
-    body: {
-      socketId,
-      command: `rm -r ${repoPath}`
+
+export default async function runScript(req) {
+    const { assignment, socketId } = req.body;
+    const { repo, input, path, bonusDate, name } = assignment;
+    const repoPath = `.private_repo/${path.match(/.*\/(.+?)\/?$/)[1]}`;
+
+    try {
+        await removeProjectDirFromPrivateRepoIfItExist(repoPath, socketId);
+        await transferFileContentToServer(socketId, input, `${repoPath}/input.txt`);
+        await copyFilesToGradeToPrivateRepo(repoPath, path, socketId);
+        const data = await getResultFromGradingScript(repoPath, bonusDate, socketId);
+        const previewList = await getHTMLOutPutFileList(name, socketId);
+
+        if (!data.graders || !previewList) {
+            return Promise.reject({ message: 'Error running script! Please make sure repository path is correct!' });
+        }
+
+        const bonusList = await getBonusList(name, socketId);
+        await saveStudentResults(data, previewList, repo, name, bonusList);
+
+        return data;
+    }catch(err) {
+        console.log(err);
+        return Promise.reject({ message: err });
     }
-  }).then(() => {
-    return transfer({
-      body: {
-        socketId,
-        content: input,
-        filePath: `${repoPath}/input.txt`
-      }
-    }).then(() => {
-      // Copy all files to directory
-      // Copy script to directory
-      // Change permission on script
-      // dos2unix equivalent on script
-      // Run script
-      return command({
+}
+
+async function removeProjectDirFromPrivateRepoIfItExist(repoPath, socketId) {
+    const bashCommand = `pwd;rm -r ${repoPath}`;
+    return executeCommandInShell(socketId, bashCommand);
+}
+
+async function copyFilesToGradeToPrivateRepo(repoPath, path, socketId) {
+    const bashCommand = `cd ${repoPath};pwd;
+                   cp -r ${path}/* .;
+                   cp ../*_script.sh .;`;
+
+    return executeCommandInShell(socketId, bashCommand);
+}
+
+async function getResultFromGradingScript(repoPath, bonusDate, socketId) {
+    const bashCommand = `cd ${repoPath};ls;
+                   chmod 777 *.sh;
+                   ./*.sh ${bonusDate}`;
+
+    let res = await executeCommandInShell(socketId, bashCommand);
+
+    if (res.indexOf('Error') >= 0)
+        return Promise.reject({ message: res });
+
+    return { graders: res ? res.split('\n') : null }
+}
+
+async function getHTMLOutPutFileList(name, socketId) {
+    const bashCommand = `cd .private_repo/${name};
+                    shopt -s nullglob;
+                    files=(*/*.out.html);
+                    printf "%s\\n" "$\{files[@]}";`;
+
+    return executeCommandInShell(socketId, bashCommand);
+}
+
+async function getBonusList(name, socketId) {
+    const bashCommand = `cd .private_repo/${name};cat bonusList;`;
```

Comment:
Separate this line so its easier to read
```
const bashCommand = `cd .private_repo/${name};
                     cat bonusList;`;
```

Reply:
done